HTTP Server: Serve Grafana with a custom URL path prefix

jan25 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our [`CONTRIBUTING.md`](https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md) guide.
2. Ensure you have added or ran the appropriate tests for your PR.
3. If it's a new feature or config option it will need a docs update. Docs are under the docs folder in repo root.
4. If the PR is unfinished, mark it as a draft PR.
5. Rebase your PR if it gets out of sync with master
6. Name your PR as `<FeatureArea>: Describe your change`. If it's a fix or feature relevant for changelog describe the user  impact in the title. The PR title is used in changelog for issues marked with `add to changelog` label. 
-->

**What this PR does / why we need it**:
Serving grafana from a non-root URL will be useful for cases of path based application routing. Currently only way to achieve this is by having a proxy with subpath rewrite rules in front of Grafana.

This PR aims to provide the subpath behavior without having to setup proxy each time we want to do path based routing.

See #16623 and #16613 for discussion around this topic.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #16623

**Special notes for your reviewer**:
<s>Marking this PR as WIP temporarily, so reviewers could give suggestions on implementation.</s>

Currently this fix is using prefix url api from macaron, which seem to work fine on manual testing.

jan25 committed:
use app subUrl with macaron prefixUrl

marefr commented:
Thanks. This will break a lot of installs using [reverse proxy with sub path](https://grafana.com/docs/installation/behind_proxy/#examples-with-sub-path-ex-http-foo-bar-grafana) so I'm afraid you'll need to introduce a new setting to handle this. At least that's the current only solution I see right now together with documentation for us to be able to accept this feature.

jan25 commented:
@marefr You're right. I'll add a new setting for subpath. Should we mark the current `root_url` subpath solution as deprecated as part of this PR? Maybe we could edit docs and add a comment in `defaults.ini` file to indicate this change.
Thank you!

jan25 committed:
add sub_path setting

torkelo commented on pkg/api/http_server.go:
```diff
@@ -227,6 +227,9 @@ func (hs *HTTPServer) addMiddlewaresAndStaticRoutes() {
 	m.Use(middleware.AddDefaultResponseHeaders())
 
+	if setting.AppSubUrl != "" {
+		m.SetURLPrefix(setting.AppSubUrl)
```
nice, did not know maracon had this, thought we would have to modify all macaron routes :)

jan25 committed:
add setting flag to use(or not use) proxy for non-root traffic

jan25 committed:
adjust docs in .ini files

jan25 committed:
remove stale fn for base path setting

jan25 commented:
Added `use_proxy_setup` bool flag setting to use proxy setup or not. Defaults to true(for backward compatibility) and to not break current installs. If set to false Grafana is served from `root_url`; and no proxy required in-front.

May need to update documentation accordingly?

marefr reviewed (changes requested):
Thanks. This looks quite good now. 

I still don't like the naming of the setting though and it should be defined so the value can be false per default - We try to enforce naming so that default boolean settings are false.

I don't have any great ideas of naming, but here they go anyway:
- `serve_from_sub_path = false`
- `serve_from_app_sub_url = false`
- `append_sub_path = false`
- `append_sub_path = false`

Thoughts/ideas?

Documentation needs to be updated with the new configuration parameter in this [file](https://github.com/grafana/grafana/blob/master/docs/sources/installation/configuration.md).

jan25 committed:
rename serve at subpath setting

jan25 committed:
update installation docs with serve_from_sub_path setting

jan25 commented:
Thanks for review @marefr . `serve_from_sub_path` sounds good, I think. Also, updated the installation docs.

And, any hints whats up with codespell checker? Have we changed something in it recently? It is [failing](https://circleci.com/gh/grafana/grafana/103770?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link) for my recent commits in unrelated files.

marefr reviewed (commented):
Think we can merge this soon, looks good. Still need to do some manual testing of some edge cases just to verify they works as expected. Until then, please merge your branch with latest master to resolve the codespell issues (I think).

marefr reviewed (commented):
Just manually tested this and it works really nice :+1: Verified some special middleware routes:
- /metrics works as expected
- /api/health works as expected

So please merge or rebase with master and we can merge this. Thanks

jan25 committed:
Merge branch 'master' into subpath-feat

jan25 commented:
CI passed after merging master. Thanks

marefr committed:
update configuration files and docs

marefr commented:
@jan25 thanks. Pushed some configuration file and docs updates. This should be ready for merge :+1:

marefr committed:
docs: update

marefr commented:
Thank you for contributing to Grafana!

kesor commented:
Works great! Thank you!

```
docker run -ti --rm --name=grafana \
    -e GF_SERVER_ROOT_URL='%(protocol)s://%(domain)s/app/grafana' \
    -e GF_SERVER_SERVE_FROM_SUB_PATH=true \
    -p 3000:3000 \
    grafana/grafana:master
```

Ghazgkull commented:
This is a great feature and I'm excited for it to be available in the next release.

One word of caution: This setting currently appears in the live configuration docs on grafana.com, although it only works when running from master. Until the feature is actually released, anyone trying to use this setting will be confused.

https://grafana.com/docs/installation/configuration/#serve-from-sub-path

marefr commented:
@Ghazgkull thanks for pointing that out. We sometime miss to add that. Would you be interested to submit a fix for this? What we want to do is add an availability note like this one https://grafana.com/docs/installation/configuration/#error-or-timeout to the serve_from_sub_path setting. Let me know, otherwise we'll try to get back to this in a while and update the docs.

Ghazgkull commented:
Sure, I can do that. Do you have any particular language you use for features that are not yet available in a release? Should we just assume that it will be available in the next minor version and say something like: `Will be available in v6.2.5 and above`?

marefr commented:
@Ghazgkull sorry. Should be: `Available in 6.3 and above`
